src/fsharp/service/ServiceAnalysis.fs

```diff
@@ -153,45 +155,65 @@ module UnusedOpens =
+        for openedModule in newlyOpenedModules do
+            let scopes =
+                match usedModules.TryGetValue openedModule.Entity with
+                | true, scopes -> openStatement.AppliedScope :: scopes
+                | _ -> [openStatement.AppliedScope]
+            usedModules.[openedModule.Entity] <- scopes
+        newlyOpenedModules.Length > 0
                                           
     /// Incrementally filter out the open statements one by one. Filter those whose contents are referred to somewhere in the symbol uses.
     /// Async to allow cancellation.
-    let rec filterOpenStatementsIncremental symbolUses (openStatements: OpenStatement list) (usedModules: UsedModule list) acc = 
+    let rec filterOpenStatementsIncremental symbolUses2 (symbolUsesRangesByDeclaringEntity: Dictionary<FSharpEntity, range list>) (openStatements: OpenStatement list)
+                                            (usedModules: Dictionary<FSharpEntity, range list>) acc = 
         async { 
             match openStatements with
             | openStatement :: rest ->
-                match getUsedModules symbolUses usedModules openStatement with
-                | [] -> 
+                if isOpenStatementUsed symbolUses2 symbolUsesRangesByDeclaringEntity usedModules openStatement then
+                    return! filterOpenStatementsIncremental symbolUses2 symbolUsesRangesByDeclaringEntity rest usedModules acc
+                else
                     // The open statement has not been used, include it in the results
-                    return! filterOpenStatementsIncremental symbolUses rest usedModules (openStatement :: acc)
-                | moreUsedModules -> 
-                    // The open statement has been used, add the modules which are already known to be used to the list of things we don't need to re-check
-                    return! filterOpenStatementsIncremental symbolUses rest (moreUsedModules @ usedModules) acc
+                    return! filterOpenStatementsIncremental symbolUses2 symbolUsesRangesByDeclaringEntity rest usedModules (openStatement :: acc)
             | [] -> return List.rev acc
         }
 
+    let entityHash = HashIdentity.FromFunctions (fun (x: FSharpEntity) -> x.GetEffectivelySameAsHash()) (fun x y -> x.IsEffectivelySameAs(y))
+
     /// Filter out the open statements whose contents are referred to somewhere in the symbol uses.
     /// Async to allow cancellation.
-    let filterOpenStatements symbolUses openStatements =
+    let filterOpenStatements (symbolUses1: FSharpSymbolUse[], symbolUses2: FSharpSymbolUse[]) openStatements =
         async {
-            let! results = filterOpenStatementsIncremental symbolUses (List.ofArray openStatements) [] []
+            let symbolUsesRangesByDeclaringEntity = Dictionary<FSharpEntity, range list>(entityHash)
+            for symbolUse in symbolUses1 do
```

Comment:
Could you add a few lines of comments around here please, e.g. briefly describe the table.